### PR TITLE
Add an anti-cache for SymbolLookup to resolve performance issues

### PR DESCRIPTION
**Pull request by svanoort**
I noted repeated, heavy ClassLoading activity tied to `StepDescriptor.metaStepsOf(String symbol)` -- this traced to `SymbolLookup.get().findDescriptor(Class a, String symbol)` -- which does an exhaustive scan apparently when getting a null hit. 

We normally can't cache misses because plugins might add new Symbol implementations... but by checking the plugin list for changes we CAN cache. 

TODO:

- [x] Some sort of test involving the anticache unloading?  I'm not even sure how or *if* we can test that in a reasonable Jenkins testcase with JenkinsRule

### CODE REVIEW

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java [hunk 1]
```diff
@@ -31,12 +34,39 @@
...
+
+    /** Update list of plugins used and purge the noHit cache if plugins have been added
+     */
+    private synchronized void checkPluginsForChangeAndRefresh() {
+        List<PluginWrapper> wrap = pluginManager.getPlugins();
+        Set<String> names = pluginsToNames(wrap);
+
+        if (wrap.size() != pluginNames.size() || !(pluginNames.containsAll(names))) {
+            this.pluginNames = names;
+            noHitCache.clear();
```
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: Deserves a test I think: create a tiny plugin with a symbol, put the `*.hpi` into `src/test/resources/`, then have the test look up that symbol, call `PluginManager.somethingOrOther` to install the plugin, look up the symbol again.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java [hunk 2]
```diff
@@ -31,12 +34,39 @@
 public class SymbolLookup {
     private final ConcurrentMap<Key,Object> cache = new ConcurrentHashMap<Key, Object>();
 
+    private final ConcurrentMap<Key,Object> noHitCache = new ConcurrentHashMap<Key, Object>();
+
+    static final Object NO_HIT = new Object();
```
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: Overly complex typing here. Since this is merely a flag, it would suffice for this to be a `Map<Key, Boolean>` (or even a `ConcurrentHashSet` if we had one of the libraries providing one handy).

I suppose you are not just folding this into `cache` (retaining `NO_HIT` as a special value) since you expect the code path for cache misses to be less often encountered than positive hits, and wish to avoid calculating the list of plugin names too often.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java [hunk 3]
```diff
@@ -31,12 +34,39 @@
...
+
     @Inject
     PluginManager pluginManager;
 
     @Inject
     Jenkins jenkins;
 
+    Set<String> pluginNames = Collections.EMPTY_SET;
```
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: `emptySet()`

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java [hunk 4]
```diff
@@ -31,12 +34,39 @@
...
+        return pluginNames;
+    }
+
+    /** Update list of plugins used and purge the noHit cache if plugins have been added
+     */
+    private synchronized void checkPluginsForChangeAndRefresh() {
+        List<PluginWrapper> wrap = pluginManager.getPlugins();
+        Set<String> names = pluginsToNames(wrap);
+
+        if (wrap.size() != pluginNames.size() || !(pluginNames.containsAll(names))) {
```
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: Is this just a roundabout way of saying

```java
if (!names.equals(pluginNames)) {
```

?

**Comment by svanoort** [comment]
@reviewbybees fwiw

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java [hunk 5]
```diff
@@ -133,13 +177,12 @@ public boolean equals(Object o) {
             if (this == o) return true;
             if (o == null || getClass() != o.getClass()) return false;
             Key key = (Key) o;
-            return type==key.type && tag.equals(key.tag) && name.equals(key.name);
-
+            return type.getName().equals(key.type.getName()) && tag.equals(key.tag) && name.equals(key.name);
```
Comment (jglick) [inline plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: Why this change? It will mean that similarly-named classes from different class loaders will compare as equal in `Key`, which seems wrong.

File: plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    /** Verify that if we install a new plugin with a Symbol use, that symbol is found.
+     *  Without the plugin we hit the "anticache" i.e. {@link SymbolLookup#noHitCache} that the symbol does not exist.
+     *  Once we add the plugin it should hit the cache.
+     */
+    @Test
+    public void testAnticache() throws Exception{
+        Descriptor d = SymbolLookup.get().findDescriptor(SimpleBuildStep.class, "trivialBuilder");
+        Descriptor stepDescriptor = SymbolLookup.get().find(Descriptor.class, "trivialBuilder");
+        assert d == null;
```
Comment (jglick) [inline plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java]:
```java
assertNull(d);
```

File: plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        assert stepDescriptor == null;
+
+        // See the test-plugin folder to rebuild this
+        File plugin = new File(SymbolLookupAnticacheTest.class.getResource("/structs-test-plugin.hpi").getFile());
```
Comment (jglick) [inline plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java]: Will break in some directory paths. Use

```java
File plugin = new File(SymbolLookupAnticacheTest.class.getResource("/structs-test-plugin.hpi").toURI());
```

File: plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        rule.jenkins.getPluginManager().dynamicLoad(plugin);
+
+        d = SymbolLookup.get().findDescriptor(SimpleBuildStep.class, "trivialBuilder");
+        stepDescriptor = SymbolLookup.get().find(Descriptor.class, "trivialBuilder");
+        assert d != null;
```
Comment (jglick) [inline plugin/src/test/java/org/jenkinsci/plugins/structs/SymbolLookupAnticacheTest.java]:
```java
assertNotNull(d);
```

File: pom.xml [hunk 9]
```diff
@@ -37,6 +37,7 @@
   <properties>
     <jenkins.version>1.580.1</jenkins.version>
     <java.level>6</java.level>
+    <java.test.level>8</java.test.level>
```
Comment (jglick) [inline pom.xml]: Do not use this property. If you are not updating the Jenkins baseline, test sources must remain at `-source 6`.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 5 above)
Comment (svanoort) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: @jglick No, that's exactly the intent -- if the full classname matches, I want to ensure we don't get an incorrect cache miss due to a nested classloader or similar.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 2 above)
Comment (svanoort) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: Correct, I also don't want to have to purge both caches.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 5 above)
Comment (jglick) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: That makes no sense that I can see. If a `Key.type` has been successfully loaded, it has a particular defining loader. Any initiating loader finding it will produce the same `Class` object. You would only have distinct classes in case the type was loaded from distinct places, such as a fork of a library, in which case the class definition could be quite different.

Was this diff line prompted by some actual experience you can point to?

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 2 above)
Comment (jglick) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: so can simplify to a `Boolean` value type

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 5 above)
Comment (svanoort) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: @jglick  I have in a few rare cases been bitten by classloading quirks like this, yes.  I do NOT agree that it introduces a defect, but I also don't care to fight this one out.

Comment (jglick) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: I just studied the sources and I believe it would introduce a `ClassCastException` under some conditions. Unlikely ones, but the `equals` code as it was originally was both correct and faster than the variant you proposed.

File: plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java (hunk 4 above)
Comment (svanoort) [reply plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java]: It's a guaranteed deep comparison - happens to be equivalent for HashSets.
